### DISCUSSION TITLE
Python wrapper for Dakota

### DISCUSSION THREAD

**Discussion by wvangeit** [snl-dakota/dakota#73] [Q&A]
Sorry if this question has been asked before, but I couldn't find anything in the history of the discussions.
I would like to use Dakota completely from within Python. Are there any internal/external (maintained) Python wrappers that would make this possible.

I'm talking about a wrapper that would allow:
- to start Dakota from within python
- manipulate the settings of the algorithms (without having to manually create  the .in files)
- access the output of the algorithms
- provide python drivers for the evaluations

Something that could already partially do any of these would be good.

I'm mostly asking because I might create something myself that can do this, but I don't want to duplicate existing work.

Thank you.

**Comment by joemar25** [comment]
Yes, there is a Python wrapper for DAKOTA called `pyDAKOTA`. This wrapper allows you to use DAKOTA completely from within Python. Here's what you can do with it:

- Start DAKOTA from within Python: You can import DAKOTA in your Python script and run it using the `run_dakota` function.
- Manipulate the settings of the algorithms: While the documentation does not explicitly mention this, it is likely that you can manipulate the settings of the algorithms through the Python interface.
- Access the output of the algorithms: The Python wrapper should allow you o access the output of the algorithms.
- Provide Python drivers for the evaluations: The wrapper allows for DAKOTA to call a Python object with a "dakota_callback" function for function evaluations.

You can find more information and the source code for `pyDAKOTA` on its [GitHub page](https://github.com/WISDEM/pyDAKOTA). Please note that you will need to install DAKOTA first before you can use `pyDAKOTA`. The installation instructions are also provided on the GitHub page.

**Reply by jadamstephens** [reply]
Thanks for posting this. I was not aware this existed!

**Reply by wvangeit** [reply]
Thank you, this is already useful. It's a bit scary that this code doesn't seem to have been touched for years, but in the worst case it can be used as basis.

**Comment by jadamstephens** [comment]
The Dakota team provides a Python binding, but it's pretty basic and doesn't have all the features you want. For one thing, it still requires a Dakota input file. You can see a test that exercises some of its features [here](https://github.com/snl-dakota/dakota/blob/devel/src/unit/test_dakota_python_env.py).

We are trying this (fiscal) year to do some refactoring that will hopefully enable more capable Python bindings, but it's not a small or simple task.

Edit: I forgot to mention that you'll need to build Dakota from source to create the binding. The CMake flag is ``DAKOTA_PYTHON_WRAPPER``.

**Reply by wvangeit** [reply]
Thank you for your answer. That's also interesting indeed.

Yes, I indeed had already recompiled Dakota with most its options enabled (which was a bit ... challenging ;-) ).

I think it would be really really useful to have dakota be integrated in python, imo it would create give a large boost to its userbase, since this is a really useful tool.

**Reply by wvangeit** [reply]
@jadamstephens Do you know if there is any news about the more tight python binding. I'm mostly interested in handling the configuration using Python. 
Your pointer to the dakota.environment examples was extremely helpful, I'm using it all the time now.
However I'm still struggling with the input configuration. Ideally I would like to feed in a JSON file or so as configuration, but I don't see a clear path to the translation from JSON to NIDR. (I saw that there are scripts to transform the XML spec to NIDR spec, but not for config files themselves).
